### PR TITLE
Introducing std.meta package

### PR DESCRIPTION
**Pull request by Dicebot**
Follow-up to https://github.com/D-Programming-Language/phobos/pull/2687

Changes from original PR:
1) have split the package in several modules: algorithm, internal, list, predicates
2) removed `static` and `template` prefixes from names now that they have own module namespace
3) moved `MostDerived` and `DerivedToFront` to `std.traits` (no idea why those added to `std.typetuple` in the first place

TODO:
 - [x] read through documentation 
 - [x] provide fine-tuned deprecation messages in `std.typetuple` for each specific symbol
 - [x] stop using `std.typetuple` in Phobos itself
 - [x] prepare DMD tests (https://github.com/D-Programming-Language/dmd/pull/4613)
 - [x] provide replacements for symbols that have `TypeTuple` as part of their name

I hope I have resisted all urges to improve / enhance actual functionality of moved templates.

(poke @JakobOvrum)

### CODE REVIEW

**Comment by John-Colvin** [comment]
Do you have something like this in here yet? https://gist.github.com/John-Colvin/2df6f4f12bfdadc10237

Or should I wait until the dust has settled?

**Comment by JakobOvrum** [comment]
@John-Colvin, https://github.com/D-Programming-Language/phobos/pull/1472

File: std/typetuple.d [hunk 1]
```diff
@@ -37,16 +40,17 @@ module std.typetuple;
 /**
  * Creates a typetuple out of a sequence of zero or more types.
  */
-template TypeTuple(TList...)
+deprecated("Renamed to std.meta.list.MetaList")
+template MetaList(TList...)
```
Comment (MartinNowak) [inline std/typetuple.d]: You replaced `TypeTuple`.

**Comment by MartinNowak** [comment]
> @MartinNowak could we possibly aim this for 2.068?

OK, let's do it, we shouldn't deprecate std.typetuple right away though.
The leftover aliases in std.typetuple should forward to std.meta and std.typetuple should be documented as going away, but deprecation would only be the next step.

**Comment by JakobOvrum** [comment]
`allSatisfy` and `anySatisfy` seem to have taken a hit to conciseness. Were just `any` and `all` ambiguous with those in std.algorithm?

Other than that, LGTM. Glad that this mostly makes template constraints *shorter*, as they had a tendency to get a bit long-winded.

**Comment by schveiguy** [comment]
> Can't reproduce the failure locally. Is it something about auto-tester?

Yes. Just deprecate the test

**Comment by schveiguy** [comment]
In the auto tester. You just deprecate that test and it will rerun

**Comment by JakobOvrum** [comment]
@andralex, want to merge? Martin is on board with aiming this for 2.068.

**Comment by JakobOvrum** [comment]
Auto-merge toggled on

**Comment by JakobOvrum** [comment]
What's with the core.sync.mutex failures?

https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=1537122&isPull=true

**Comment by JakobOvrum** [comment]
I think I've seen it happen for other PRs too, so it's probably not caused by this PR. I'm not sure if it's meaningful or not to deprecate those tests, though, we need to figure out what caused it.

**Comment by MartinNowak** [comment]
Not related to your pull, we have a huge increase of spuriously failing tests recently (Linux 32). Didn't yet figure out what's causing this b/c I couldn't reproduce it. Maybe @braddr can catch one of those stucked runs and give us a gdb backtrace.

**Comment by andralex** [comment]
I'll start with the punchline: We need to unpull this and then implement small parts of it with extreme care. It is extremely disruptive whilst materially improving matters only very little. Details follow.

As a note related to scope and impact, I should say that this work is perfectly meaningful as a pass through a project with an eye for improving names of things. However, at this point we're very far from that stage and we need to be a lot more conservative than regularly going through the names in look for the elusive perfect naming scheme.

Generally any name change we make should be only for very good reasons (e.g. the old name was mightily confusing - e.g. `TypeTuple`) and also keep the old name with minimal documentation so people who search with google can find it. For example: "`TypeTuple` is an alias for `MetaList`, kept for background compatibility reasons. New code should use the more descriptive `MetaList`."

Point by point:

* isExpressionTuple -> isExpressionList does not add enough value
* TypeTuple -> MetaList does add value, replace and keep the old one as an alias (this is pretty much the main change I agreed with in the past)
* std.typetuple -> std.meta is valuable, but keep the former to just import std.meta. Remove from the modules list but mention it in the docs of std.meta.
* ParameterTypeTuple -> ParameterList does not add enough value
* staticMap -> Map does not add enough value
* staticIndexOf -> indexOf does not add enough value
* allSatisfy -> all does not add enough value
* anySatisfy -> any does not add enough value
* FieldTypeTuple -> FieldTypes does not add enough value
* The breakage of std.meta née std.typetuple into a package seems just busywork to me. It would need to be better motivated.
* Don't deprecate MostDerived and DerivedToFront, just silently alias them across modules.
* Most if not all renamings in std/traits.d (ParameterDefaultValueTuple -> ParameterDefaultValues etc) do not add enough value.

Finally allow me an appeal to cooler heads and good judgment. Please, please, please let's not make this yet another flamewar and protracted negotiation. I understand many of the newer names are better, but other schemes will be better schemes. We can't afford this disruption. This is not a semester science project and we must stop treating it as one. Thanks.

**Comment by andralex** [comment]
Forgot to mention - std.typelist should stay, in the worst case relegated to a scarcely documented stub.

**Comment by MartinNowak** [comment]
Thanks Andrei for the valuable comment. We pulled this based on the discussion in #2687, but there were little if any arguments against renaming, and the fact that renaming things has a cost isn't yet internalized, so we'll have to explicitly say that a few times more.
That said, one of the main goals of this pull is to replace TypeTuple with a clearer name, so we should also rename ParameterTypeTuple and FieldTypeTuple.

**Comment by jacob-carlborg** [comment]
I don't see why `staticMap` was renamed to `Map` with a capital "M" but `anySatisfy` was renamed to `any` with a lowercase "a".

**Comment by JakobOvrum** [comment]
> Generally any name change we make should be only for very good reasons (e.g. the old name was mightily confusing - e.g. TypeTuple) and also keep the old name with minimal documentation so people who search with google can find it. For example: "TypeTuple is an alias for MetaList, kept for background compatibility reasons. New code should use the more descriptive MetaList."

The name `TypeTuple` doesn't confuse people because of the `Type` part, it's the `Tuple` part. All the other renamings are done in exactly the same vein.

`TypeTuple` itself is actually a relatively minor issue since it's hardly ever seen in public interfaces. If we leave references to "tuple" in the names of templates that return meta lists then this whole thing is moot.

The reason I've been backing this from the start is because this is one of the things that I have to explain to newbies again and again, but apparently we cannot afford the disruption. Exactly what disruption does this cause? It doesn't break any code.

> I don't see why `staticMap` was renamed to `Map` with a capital "M" but `anySatisfy` was renamed to any with a lowercase "a".

`any` always evaluates to a value (boolean), while `Map` evaluates to a list that may contain both types and values.

**Comment by John-Colvin** [comment]
> Generally any name change we make should be only for very good reasons (e.g. the old name was mightily confusing - e.g. TypeTuple) and also keep the old name with minimal documentation so people who search with google can find it. For example: "TypeTuple is an alias for MetaList, kept for background compatibility reasons. New code should use the more descriptive MetaList.

Tuple needs to be entirely erased (aliases for backwards compatibility excepted) from phobos unless it's referring to std.typecons.Tuple. It's not about finding the best names, which I agree is inappropriate at this stage, but about having the slightest bit of consistency and making it explainable, as opposed to "oh that's just the other thing we also call tuple, it's kinda similar sometimes but not really" which is a huge put-off for new users coming to enjoy what is after all a big selling point for D: it's metaprogramming.

You are right, we shouldn't endlessly churn looking for better names, but this is not one of those cases. The current situation is horrendous, this pull makes it liveable.

> The breakage of std.meta née std.typetuple into a package seems just busywork to me. It would need to be better motivated.

The split provides a setting for adding a bunch more advanced and/or convenience metaprogramming tools. This pull deliberately doesn't add any of these, but they are there waiting in the wings. There is enough out there to make manipulating MetaLists almost as easy as using ranges, but it shouldn't all be dumped in one enormous module.

**Comment by andralex** [comment]
>The name TypeTuple doesn't confuse people because of the Type part, it's the Tuple part. 

I won't argue this much, but I question this assertion. Per m-w.com: "set of (so many) elements —usually used of sets with ordered elements". The trouble is with Type because a TypeTuple can contain things that were emphatically not types.

If you convinced me it wasn't the Type part that was confusing, then I'd be against any renaming whatsoever.

>That said, one of the main goals of this pull is to replace TypeTuple with a clearer name, so we should also rename ParameterTypeTuple and FieldTypeTuple.

TypeTuple barely makes the bar for renaming, the others I don't think do. Please let's not turn this into a negotiation. I said and I repeat: some of the new names are clearly better but not enough. There's no intrinsic value in renaming ParameterTypeTuple to Parameters - either way it has to have a name that has to be found by perusing the documentation. It's not something natural that can be inferred without any lookup. Tomorrow someone comes with the `Pack` convention, inspired from C++. So we'll have everyone mesmerized on `ParameterPack` and `FieldTypePack` etc. Please let's stop. The quest for slightly better names could go on forever.

**Comment by JakobOvrum** [comment]
> I won't argue this much, but I question this assertion. Per m-w.com: "set of (so many) elements —usually used of sets with ordered elements". The trouble is with Type because a TypeTuple can contain things that were emphatically not types.

I agree with the logic, but it's irrelevant - the issue is that users confuse std.typecons.Tuple with template argument lists.

> If you convinced me it wasn't the Type part that was confusing, then I'd be against any renaming whatsoever.

The issue has always been the name "tuple". Feel free to dig up the old threads and verify.

> TypeTuple barely makes the bar for renaming, the others I don't think do. Please let's not turn this into a negotiation. I said and I repeat: some of the new names are clearly better but not enough. There's no intrinsic value in renaming ParameterTypeTuple to Parameters - either way it has to have a name that has to be found by perusing the documentation. It's not something natural that can be inferred without any lookup. Tomorrow someone comes with the Pack convention, inspired from C++. So we'll have everyone mesmerized on ParameterPack and FieldTypePack etc. Please let's stop. The quest for slightly better names could go on forever.

This ignores the root issue that is the inappropriate overloading of the concept "tuple" in the D ecosystem.

**Comment by andralex** [comment]
>Tuple needs to be entirely erased (aliases for backwards compatibility excepted) from phobos unless it's referring to std.typecons.Tuple.

I disagree. I don't see a problem acknowledging there are tuples of values and tuples of other things. By the same argument, "List" should be removed unless it refers to a list of values - wait a minute, this very diff has a `MetaList`!

Please stop.

> It's not about finding the best names, which I agree is inappropriate at this stage, but about having the slightest bit of consistency and making it explainable, as opposed to "oh that's just the other thing we also call tuple, it's kinda similar sometimes but not really" which is a huge put-off for new users coming to enjoy what is after all a big selling point for D: it's metaprogramming.

Breaking 50+ modules of Phobos doesn't seem quite inspiring for those folks who've been using metaprogramming and will find they need to change names in random ways. 

>You are right, we shouldn't endlessly churn looking for better names, but this is not one of those cases. The current situation is horrendous, this pull makes it liveable.

Yeah, it's always different this one time, or it's the last time.

Please do not. We can't afford all this anymore.

**Comment by andralex** [comment]
>This ignores the root issue that is the inappropriate overloading of the concept "tuple" in the D ecosystem.

If that's the root issue here, we can sleep soundly.

**Comment by andralex** [comment]
> there were little if any arguments against renaming, and the fact that renaming things has a cost isn't yet internalized, so we'll have to explicitly say that a few times more.

Wisest words in this thread. Thanks @MartinNowak

**Comment by JakobOvrum** [comment]
> If that's the root issue here, we can sleep soundly.

You're free to dismiss the problems newcomers face at your own disadvantage as co-head of D, but at what gain? This doesn't break any code and it doesn't ever need to.

Even the [DIP](http://wiki.dlang.org/DIP54) states the problem quite clearly in the rationale and description. You've been participating in the process all along, you don't get to claim to be ignorant of the rationale and shoot it down now on that basis.

> there were little if any arguments against renaming, and the fact that renaming things has a cost isn't yet internalized, so we'll have to explicitly say that a few times more.

This was always in the back of everyone's heads. See comments like [this one](https://github.com/D-Programming-Language/phobos/pull/2687#issuecomment-63125156).

Breaking code doesn't further the cause in any way. What we need is a distinction between the two "tuples" so it's actually possible to teach these things to beginners. The cost of this is two-fold; 1) all the work that people like @Dicebot had to do to fix the situation which has already been paid, and 2) the trivial cost of maintaining the old symbols for compatability.

**Comment by andralex** [comment]
>You're free to dismiss the problems newcomers face at your own disadvantage as co-head of D, but at what gain? This doesn't break any code and it doesn't ever need to.

I thought I was dismissing a problem _you_ assert to exist.

>Even the DIP states the problem quite clearly in the rationale and description. You've been participating in the process all along, you don't get to claim to be ignorant of the rationale and shoot it down now on that basis.

If I participated in the discussion, it doesn't mean I automatically must accept whatever comes out of it.

This is not shooting down and please don't attempt to make it an ethical matter for me. I'm all for good work that makes Phobos better. I won't accept code because it has a DIP that's been discussed and because it's extensive. It is Walter's and my assessment that this work does not improve the lot of current and future D users.

**Comment by WalterBright** [comment]
> This doesn't break any code

Yes, it does. std.typelist is removed and several declarations are deprecated in std.typetuple. Furthermore, these renamings obsolete books, articles, and tutorials written about D. And lastly, there have been several of these grand renamings, each billed as the last one.

> all the work that people like @Dicebot had to do to fix the situation which has already been paid

I find this as upsetting as anyone, as we cannot squander effort. But adopting something because a lot of work was put into it, rather than on its merits, is the sunk cost fallacy: http://en.wikipedia.org/wiki/Sunk_costs

**Comment by JakobOvrum** [comment]
> I thought I was dismissing a problem you assert to exist.

If that's the case, then we're back to square one. I'd like to forget about this massive waste of time as soon as possible.

**Comment by WalterBright** [comment]
TypeTuple is clearly a misleading name, and is causing ongoing problems. But using that as justification for a grand renaming is a costly overreach. And changing `Tuple` to `List` has misleading implications, too, as a list implies being a linked list, as opposed to the array-like behavior that tuples have. In my own code I've gravitated for names like `Parameters` for an aggregation of `Parameter`s. It doesn't imply what sort of container it is, it's generic, and hits the sweet spot (obviously, I'm not a fan of Hungarian notation).
Since pretty much all of Phobos is going meta, having a new package named `meta` is not going to be very helpful moving forward.
Creating new packages, new modules, removing old ones, moving names from one module to another, etc., is not nearly enough value for the cost. So here is what we're going to do:

1. There's going to be zero code breakage. Old names will stay as documented aliases to the new names, so that old code will continue to work and that googling the names will work. It is important to not break books, articles, tutorials, etc., and the uncounted investment in them.

2. Existing modules stay and will not be renamed.

3. Names will not move from one module to another.

4. The following renames will happen:
isExpressionTuple => isExpressions
ParameterTypeTuple => Parameters
FieldTypeTuple => Fields
ParameterDefaultValueTuple => ParameterDefaults
TypeTuple => Arguments

This is the least disruptive change while preserving the value.

**Comment by John-Colvin** [comment]
@WalterBright that sounds reasonable, with one exception:

std.meta is a package for manipulating and creating compile-time objects. This isn't the same as the general templatisation of phobos. If you really think it's not worth it then so be it (I strongly disagree here), but that's not a valid reason.

**Comment by WalterBright** [comment]
@dicebot As you say, many of D's constructs have unique characteristics and so using other language semantics as a guide has only limited utility. D tuples, for example, can be indexed which is not an implied characteristic of a list. Hence, renaming them to lists is a step sideways. Your idea of simply appending `s` is a superior idea.

@John-Colvin Everything in Phobos is going meta. Any package called meta is just going to become another "scrapyard" (to borrow Dicebot's term). It's not just templates, there'll be a lot of meta stuff that is specific to what a particular module is about. For example, I don't see value in moving a meta template that is specific to strings into `meta`. It should be in `string`.

It all reminds me of my doomed attempts at categorizing my book collection (5000 books) on my computer. Do I organize them by subject, author, genre, quality, already read, age? All books fit nicely into multiple genres. There is no solution. There's no hierarchy of Phobos modules either that is going to work - the only thing we can do is just take a stab at it an provide better search/discovery methods. Trying to deal with this by having more modules that merely contain lists of aliases is not solving the problem, either, it just adds to the burden of anyone trying to understand someone else's code.

```
What’s in a name? that which we call a rose  
By any other name would smell as sweet; 
```
I'll reiterate that these endless quests for the perfect names and taxonomy is, in the end, a waste of time and destructive of D by adding cruft, baggage, and impeding communication between groups of programmers. We've had grand renamings in the past and I have yet to see a single comment of the form "Phew! what a relief that was done!" from any who didn't have a dog in the hunt, nor any evidence of improved productivity. But I have seen an awful lot of "D has never left beta", "D won't compile last year's code", "D is unstable", "D's developers cannot make up their minds", and the worst: "D's real problems are not getting resolved."

None of these renaming proposals are going to make code run faster, be smaller, compile faster, or be more correct. All of them consume vast amounts of our time, and an uncounted amount of time for the silent majority that use D out in the field. Criminy, this proposal required FIFTY modules in Phobos to be changed. FIFTY. This is not going to happen. We must be far more modest about changing a name. What I listed above is the minimum that resolves the issue, and that's why it's the solution we're going with.

Revert this, do the renamings I enumerated above, and let's move on.

**Comment by WalterBright** [comment]
> I wonder if you have noticed that this PR breaks no code. At all.

std.typelist was deleted, and std.typetuple is deprecating several names. That's breakage, as well as breaking books, articles, and tutorials.

> We had huge amount of debate and all suggestions felt bad.

My feeling as well. I know you have invested significant time in this, and I'm sorry about that. I'm sorry that we haven't been able to reach consensus, but someone has to make a decision, and that's me.

**Comment by John-Colvin** [comment]
> @John-Colvin Everything in Phobos is going meta. Any package called meta is just going to become another "scrapyard" (to borrow Dicebot's term). It's not just templates, there'll be a lot of meta stuff that is specific to what a particular module is about. For example, I don't see value in moving a meta template that is specific to strings into meta. It should be in string.

@WalterBright I'm not sure I understand you. I am talking about things like Zip, Filter, Sort, CartesianProduct and so on that work with what you are referring to as Arguments. It's not about using std.meta as a dumping ground for every specific piece of metaprogramming used in phobos, it's about providing a set of general tools, like std.algorithm/std.range do for ranges.

**Comment by MetaLang** [comment]
I'm not invested in this argument over naming in any way, but I'm curious as to how marking something deprecated constitutes a code breakage. Isn't the whole point of deprecated() that it discourages the use of a symbol without breaking code? Even then, I **thought** there was a well-defined path to deprecation, such that std.typetuple would not be marked as deprecated outright, but 6 months down the road.

**Comment by WalterBright** [comment]
> but I'm curious as to how marking something deprecated constitutes a code breakage. Isn't the whole point of deprecated() that it discourages the use of a symbol without breaking code? Even then, I thought there was a well-defined path to deprecation, such that std.typetuple would not be marked as deprecated outright, but 6 months down the road.

I think you answered the question - deprecation does break code, it just does it in a gradual, nicer manner, rather than abruptly. It still forces people, sooner or later, to go and edit their code because it won't compile anymore. It still causes working libraries that aren't actively maintained to break. It still means that an older project pulled from somewhere won't compile anymore. It still means that books that have examples using TypeTuple become obsolete.

**Comment by MetaLang** [comment]
It won't break code unless you're using the -de compiler switch or the symbols are eventually removed. `deprecated` on its own does not cause code breakage.

**Comment by WalterBright** [comment]
Selectively overlooking deprecations is not a generally practical strategy for users, and we should not be advocating it. They're all or nothing.

**Comment by WalterBright** [comment]
https://github.com/D-Programming-Language/phobos/pull/3264

**Pull request merged**